### DISCUSSION TITLE
Does module.pyi work for code in module.py?

### DISCUSSION THREAD

**Discussion by prim** [microsoft/pyright#7658] [Q&A]
I have three python file: `myscript.py` `some/module.py` and `some/module.pyi`。
type hints stored in `some/module.pyi` file。

```text

$ cat myscript.py
from typing import reveal_type
from some.module import add

c = add(1, 2)
reveal_type(c)


$ cat some/module.py
from typing import reveal_type

def add(a, b):
    return a + b

def f():
    c = add(1,"")
    reveal_type(c)


$ cat some/module.pyi
def add(a: int, b: int) -> int: ...
```

type hints in pyi file works on code outside of module.py. The variable c can be deduced to be of the correct type.

```
$ pyright myscript.py
f:\temp\mk2\myscript.py
  f:\temp\mk2\myscript.py:5:13 - information: Type of "c" is "int"
0 errors, 0 warnings, 1 information
```

type hints in pyi file does not have an effect on the code in module.py itself. The variable c can not be deduced to be of the correct type.

```text
$ pyright some/module.py
f:\temp\mk2\some\module.py
  f:\temp\mk2\some\module.py:8:17 - information: Type of "c" is "Unknown"
0 errors, 0 warnings, 1 information
```

Is there any way to make the type hint in module.pyi have an effect on the code inside module.py as well?

**Comment by erictraut** [comment]
Yes, this is how type stub (".pyi") files work. They are meant to be stand-ins for library modules that are lacking type information or are written in a language other than Python. If you are the owner of the ".py" file and would like a static code analyzer like pyright or mypy to analyzer the source code in that file, you should add the type information directly inline. Do not put the type information in a separate stub file in that case. The stub file will be used only if the module is imported by a different file.
